core/devdoc/on-out-process-gateway-modules.md

```diff
@@ -0,0 +1,484 @@
+    the module host process. This is a required argument. The identifier is any 
+    string that uniquely identifies the control channel. Duplicate strings will 
+    cause the module creation to fail.
+
+-    **Message ID**
+
+    The proxy module will send and receive gateway messages on sockets attached 
+    to this ID. This is optional in the module's JSON configuration. The 
+    identifier must be a unique string that identifies the message channel. 
+    Duplicate strings will cause the module creation to fail. If the 
+    ID is not set, the message channel URI will be composed from a UUID.
+
+-   **Activation Type**
+
+    This is an enumeration with values indicating how the hosting process will
+    be activated. It could indicate one of the following possible values:
+
+    -   **None:** An activation type of **None** indicates that the proxy module
+        does not have to do anything in particular to *activate* the module. The
+        expectation is that out-of-band measures will be adopted to ensure that
+        the out-proc module will be activated.
+
+    -   **Fork:** An activation type of **Fork** means that the proxy module
+        will attempt to fork the hosting process when the module is initialized.
+        The proxy will wait for a fixed duration in order to let the outprocess
+        process complete initialization and start listening on it’s `NN_RESP`
+        socket. Additional context information is supplied as part of the
+        configuration:
+
+        -   *Host Path*: Path to the executable that will host the module in
+            question
+        -   *Arguments*: array of command line arguments, if any.
+
+    -   **Service:** This activation type indicates that the module host is a
+        service and the proxy is expected to *start* the service during
+        initialization. The configuration might indicate whether this is an
+        *upstart* or *systemd* or a Windows system service.
+
+    -   **Container:** This activation type indicates that the out-proc module
+        is hosted in a Docker container and needs to be started during
+        initialization.
+
+-   **Module Configuration**
+
+    The "args" of the module entry is serialized as a JSON for transmission on 
+    the control channel. Once handed off to the module host, this module 
+    configuration is expected to be passed to the 
+    Module_ParseConfigurationFromJson function of the MODULE_API.
+
+The first **activation type** to implement is **None**.
+
+The proxy module will create a *nanomsg* `NN_PAIR` socket and listen for 
+gateway messages on it. It will also create an `NN_REQ` socket and connect to 
+the control channel. Once a connection with the module host process has been 
+established communication will commence as explained later in this document.
+
+Module Host
+-----------
+
+The module host is a function whose primary responsibility 
```

Comment:
I was looking for a more generic term.  We've settled on "proxy gateway" so that's the term I will use.